### ISSUE TITLE
Hard Navigation On Client side Navigation (using Link component) in Nextjs 15

### ISSUE THREAD

**Issue by NorouziM** [vercel/next.js#72383]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/musing-lalande-js9nmq?file=%2Fapp%2Fpage.tsx%3A5%2C1

### To Reproduce

	1.	Create a new page in the starter project.
	2.	Add a link to that page on the home page.
	3.	Run the app locally.
	4.	Click on the link.
	5.	A hard navigation occurs, causing the window to reload and the page document to be downloaded.

In Next.js v14, this behavior wouldn’t occur. Navigations were handled as soft navigations, resulting in the document not being downloaded on every navigation and requested only once.

### Current vs. Expected behavior

In Next.js v15, clicking on internal links leads to full page reloads (hard navigations), resulting in the entire page document being downloaded each time. This behavior differs from Next.js v14, where navigations were handled as soft navigations (client-side transitions), allowing for faster and more efficient page changes without reloading the entire document.

### Provide environment information

```bash
Operating System:
  Platform: macos
  Arch: x64
Binaries:
  Node: 20.9.0
  npm: 10.1.0
  Yarn: 1.22.19
  pnpm: 9.9.0
Relevant Packages:
  next: 15.0.2
  react: 18.3.0,
  react-dom": 18.3.0,
  typescript: 5.1.6
```

### Which area(s) are affected? (Select all that apply)

Navigation

### Which stage(s) are affected? (Select all that apply)

next dev (local), next start (local), Other (Deployed)

**Comment by james-han** [comment]
@NorouziM 
Hi NorouziM,

Thanks for bringing this up. I tested this on my Mac with Next.js v15 and noticed the same issue you mentioned—internal links were causing a full page reload. After checking the file structure, I realized the `app/x` directory was empty, so navigating to `/x` triggered a hard reload as expected. **Once I added a `page.tsx` file under `app/x`,** the reload stopped, and the navigation returned to a client-side transition.

page.tsx：

`export default function Page() {
  return <p>X Page</p>;
}`

In Next.js, a full page reload (hard navigation) can happen in these cases:
1. Missing `page.tsx`**: If the target route folder lacks a `page.tsx` file, Next.js won’t recognize it as a fully client-side route.
2. External Links**: Navigating to an external domain URL.
3. Custom Error Boundaries**: Certain errors or boundary components may cause a page refresh to reset the state.
4. Unhandled Server-only Components**: Using server-only components in client-side code or routing logic can trigger a reload.

**Comment by NorouziM** [comment]
@james-han 
Hi James-han
Thank you for your response. I wanted to clarify that I have already added the page.tsx file; I’ve attached a screenshot to confirm this. Additionally, we’re experiencing this issue on a project that was previously functioning correctly on Next.js v14. The problem only arose after upgrading Next.js to the latest version.

![image](https://github.com/user-attachments/assets/ef4f23d5-6be6-4c65-848e-b64b13ad53a1)

I’ve also included a screenshot of the simple link we’re using and a video to demonstrate the issue more clearly on the main post of this issue

**Comment by NorouziM** [comment]
@timneutkens @leerob  Any idea for this problem?

**Comment by victormanuelfrancodev** [comment]
I have the same issue but with next 14.1.0

**Comment by yarikpetrenko** [comment]
@NorouziM You have issues because with Next.js 15 cache behavior was updated and now pages are not longer cached by default see Next.js release notes - https://nextjs.org/blog/next-15#caching-semantics. So because it's not cached it requests the page each time you navigate even on the client. New configuration was added called `staleTimes ` - https://nextjs.org/docs/app/api-reference/config/next-config-js/staleTimes. That allows you to opt out to old behavior.

My question is will it be ever possible to leave dynamic behavior for some pages and have staleTimes on per page basis, because rn I feels a bit radical to have either cache for all pages or no page cache at all.

**Comment by Badestrand** [comment]
Can't believe that Next.js has this problem. React gave us client-side navigation to prevent full page-loads after each click and Next is taking it away and calls it an improvement.

**Comment by Onurfesci** [comment]
I'm experiencing the same thing, next.js seemingly randomly does hard navigation causing a full page reload. I'm on next 15.1.6 as well.

Most of the time, there's no error, sometimes there is the RSC error. This happens when I'm using `router.push`. I've tried `router.prefetch` to mitigate it but it still happens intermittently.

If there's an RSC error, it's:
```
1517-4181b53f1bb2d603.js:1 Failed to fetch RSC payload for http://localhost:3000/register/account-type/choose. Falling back to browser navigation. TypeError: Cannot read properties of undefined (reading 'call')
    at r (webpack-310e74e27d689573.js:1:143)
```

In the Network, there are 2 rsc requests then a full document request
<img width="1073" alt="Image" src="https://github.com/user-attachments/assets/253a4d84-82ae-4665-831d-63643c2ccb3a" />

Interestingly, when I do "Disable cache" in the chrome Network tab, it works perfectly - no hard navigation.

Unfortunately this has completely broken one of our step-by-step forms that we had nicely separated into logical routes, assuming next will handle everything on its end natively with navigation, which I'll have to fully refactor if this isn't resolved, along with many other parts of my app. :/

If anyone has managed to get around this please share 🙏

**Comment by RobinClowers** [comment]
> You have issues because with Next.js 15 cache behavior was updated and now pages are not longer cached by default 

@yarikpetrenko that's true, but not having a cached page doesn't cause hard navigation, otherwise the first time you visited any link it would hard navigate.

I've also noticed this issue, but for us it appears to be caused by new deploys on Vercel. I think this is by design, though I can't find it explicitly documented anywhere.
